### ISSUE TITLE
A short read on GAN training stabilization issue?

### ISSUE THREAD

**Issue by d-h-lee** [AIML-K/bamboo-forest#124]
Zhang, Hongyi, Moustapha Cisse, Yann N. Dauphin, and David Lopez-Paz. 2018. “Mixup: Beyond Empirical Risk Minimization.” In International Conference on Learning Representations. https://openreview.net/forum?id=r1Ddp1-Rb.

The key contribution of the paper is:
<img width="870" alt="image" src="https://github.com/AIML-K/bamboo-forest/assets/55645287/24320fc0-ead0-4fb3-9e86-8151617765d6">

I noticed an interesting element in this paper, regarding GAN (as many lab members are training GANs currently):
<img width="868" alt="image" src="https://github.com/AIML-K/bamboo-forest/assets/55645287/85168500-1a25-452b-8106-f5aa19c7eb15">

**Comment by wheresmadog** [comment]
Following paper regarding mixup: https://arxiv.org/abs/2010.04819

**Comment by invigorator96** [comment]
The Mixup algorithm looks similar to [This](https://github.com/AIML-K/earthquake-imputation/blob/78f55a7735b053a228ba20dcc9c6b78e2fe62a15/src/models/bbgan/bbgan_losses.py#L47-L53). However, the loss computation is quite different.

**Comment by wheresmadog** [comment]
@invigorator96 the link broken

**Comment by invigorator96** [comment]
> @invigorator96 the link broken

My mistake. It would be the permission issue. Here is the referenced code snippet:

```py
def disc_gp_loss(d_model,real_batch,fake_batch):

    '''
    d_model: BBDisc model
    real_batch: data from dataloader
    fake_batch: output of g_model, tuple of (waveforms,norm_const)
    '''
    batch_size=real_batch['x'].shape[0]

    alpha=torch.rand(batch_size).to(device='cuda')

    real_waveforms, real_norm_const = real_batch['x'], real_batch['ln_cns']
    cond_vars=real_batch['cond_var']

    fake_waveforms, fake_norm_const = fake_batch

    interpolation_waveforms= (real_waveforms * alpha[:,None,None] + fake_waveforms*(1-alpha)[:,None,None]).requires_grad_(True)
    
    interpolation_norm_const=(real_norm_const*alpha[:,None] + fake_norm_const*(1-alpha)[:,None]).requires_grad_(True)

    D_score_interpolation=d_model(interpolation_waveforms,interpolation_norm_const,cond_vars)

    Xout_wf=Variable(Tensor(batch_size,1).fill_(1.0), requires_grad=False)
    grads_wf = torch.autograd.grad(
                outputs=D_score_interpolation,
                inputs=interpolation_waveforms,
                grad_outputs=Xout_wf,
                create_graph=True,
                retain_graph=True,
                only_inputs=True,
            )[0]
    grads_wf=grads_wf.view(batch_size,-1)

    Xout_cn=Variable(Tensor(batch_size,1).fill_(1.0), requires_grad=False)
    grads_cn = torch.autograd.grad(
                outputs=D_score_interpolation,
                inputs=interpolation_norm_const,
                grad_outputs=Xout_cn,
                create_graph=True,
                retain_graph=True,
                only_inputs=True,
            )[0]
    grads = torch.cat([grads_wf, grads_cn,], 1)
    return ((grads.norm(2, dim=1) - 1) ** 2).sum()
```
